src/enclave/enclave_init.c

```diff
@@ -39,24 +39,40 @@ static void __sgxlkl_enclave_copy_app_config(
     sgxlkl_app_config_t* app_config,
     const sgxlkl_config_t* sgxlkl_config)
 {
-    int i = 0, j = 0;
+    int i = 0, envc = 0;
     char** envp = NULL;
+    size_t total_size = 0;
 
     app_config->argc = sgxlkl_config->argc;
-    app_config->argv = malloc((app_config->argc + 1) * sizeof(char*));
 
     for (i = 0; i < app_config->argc; i++)
-        app_config->argv[i] = strdup(sgxlkl_config->argv[i]);
-    app_config->argv[i] = NULL;
+        total_size += strlen(sgxlkl_config->argv[i]) + 1;
 
     envp = sgxlkl_config->argv + sgxlkl_config->argc + 1;
-    for (; envp[j++] != NULL;)
-        ;
+    while (envp[envc] != NULL)
+    {
+        total_size += strlen(envp[envc]) + 1;
+        envc++;
+    }
 
-    app_config->envp = malloc((j + 1) * sizeof(char*));
-    for (j = 0; envp[j] != NULL; j++)
-        app_config->envp[j] = strdup(envp[j]);
-    app_config->envp[j] = NULL;
+    char* buf = malloc((total_size + 1) * sizeof(char*));
+    app_config->argv = malloc((sgxlkl_config->argc + envc + 2) * sizeof(char*));
```

Comment:
It's not clear to me that this is sufficient.  We should probably mirror the layout mandated by the amd64 ELF psABI here (ideally, we'd have that structure on the stack on userspace process start, because that's what the ABI actually requires).  The layout there is (from the lowest allocated stack address upwards):

- 64-bit `argc`
- `argv` as an array of 64-bit pointers.
- 64-bit 0
- Environment pointers as an array of 64-bit pointers.
- 64-bit 0
- ELF auxiliary argument vector (array of pairs of 64-bit values)
- 64-bit 0.
- Platform-specific things, including all of the things that the pointer arrays point to.

If we have software that depends on this exact layout, it will remain broken with your code.

Reply:
Please open a PR for this and mark this as resolved.

Reply:
Which "this" is this? I assume you mean merge the current PR and open a new PR/issue for full ELF support?